### PR TITLE
[oneDNN] [tfg] [remapper] Add Conv2D + BiasAdd fusion in MLIR

### PR DESCRIPTION
**Pull request by kanvi-nervana**
This PR enables the fusion of Conv2D and BiasAdd moving Squeeze node after BiasAdd.

- Pattern before fusion Conv2D -> Squeeze -> BiasAdd
- Pattern after fusion _FusedConv2D -> Squeeze

### CODE REVIEW

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Review by rdzhabarov — APPROVED** [review]
Change LGTM! Thank you.

Looks like you have to rebase and resolve conflicts.

File: tensorflow/core/transforms/remapper/remapping_helper.h [hunk 1]
```diff
@@ -197,6 +203,11 @@ class OpPropertyHelper : public OpCatHelper {
   // Currently GPU does not supprt contraction + bias_add
   bool IsGpuCompatible(const ContractionBiasAdd&) const { return false; }
 
+  // Currently GPU does not supprt contraction + squeeze + bias_add
```
Comment (rdzhabarov) [inline tensorflow/core/transforms/remapper/remapping_helper.h:206]: nit: support.

**Comment by kanvi-nervana** [comment]
@rdzhabarov Thanks for the review. I have addressed the comment and resolved conflicts.

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Review by rdzhabarov — APPROVED** [review]

**Comment by gbaned** [comment]
Hi @kanvi-nervana Can you please rebase and resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @rdzhabarov Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk  Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by gbaned** [comment]
Hi @qingyunqu Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk, Can you please review this PR ? Thank you!

**Comment by keerthanakadiri** [comment]
Hi @penpornk, Can you please review this PR ? Thank you!

File: tensorflow/core/transforms/remapper/tests/contraction.mlir [hunk 2]
```diff
@@ -44,3 +44,20 @@ tfg.func @matmul_test() {
...
+  %Const, %ctl_1 = Const device("/device:CPU:0") name("Const") {dtype = f32, value = dense<[[[[1.11986792, -3.0272491]]]]> : tensor<1x1x1x2xf32>} : () -> (tensor<*xf32>)
+  // CHECK: %[[BIAS:.*]], {{.*}} name("Const_1")
+  %Const_1, %ctl_3 = Const device("/device:CPU:0") name("Const_1") {dtype = f32, value = dense<[0.531091094, -0.719168067]> : tensor<2xf32>} : () -> (tensor<*xf32>)
+  // CHECK: Squeeze({{.*}}) {{.*}} name("BiasAdd")
+  %Conv2D, %ctl_4 = Conv2D(%Placeholder, %Const) device("/device:CPU:0") name("Conv2D") {T = f32, data_format = "NHWC", dilations = [1, 1, 1, 1], explicit_paddings = [], padding = "SAME", strides = [1, 1, 1, 1], use_cudnn_on_gpu = true} : (tensor<*xf32>, tensor<*xf32>) -> (tensor<*xf32>)
+  %Squeeze, %ctl_5 = Squeeze(%Conv2D) device("/device:CPU:0") name("Squeeze") {T = f32, squeeze_dims = [2]} : (tensor<*xf32>) -> (tensor<*xf32>)
+  // CHECK: _FusedConv2D(%[[PLACEHOLDER]], %[[FILTER]], %[[BIAS]]) {{.*}} name("Conv2D") {{.*}} fused_ops = ["BiasAdd"]
+  %BiasAdd, %ctl_6 = BiasAdd(%Squeeze, %Const_1) device("/device:CPU:0") name("BiasAdd") {T = f32, data_format = "NHWC"} : (tensor<*xf32>, tensor<*xf32>) -> (tensor<*xf32>)
```
Comment (sueannioanis) [inline tensorflow/core/transforms/remapper/tests/contraction.mlir:61]:
```
// -----
// CHECK-LABEL: tfg.func @conv2dsqueezebias_test
tfg.func @conv2dsqueezebias_test() {
  // Define the input tensor placeholder
  %input_tensor, %ctl_input = Placeholder device("/device:CPU:0") name("input_tensor") {dtype = f32, shape = #tf_type.shape<1x3x3x1>} : () -> (tensor<*xf32>)
  
  // Define the filter tensor
  %filter, %ctl_filter = Const device("/device:CPU:0") name("filter") {dtype = f32, value = dense<[[[[1.11986792, -3.0272491]]]]> : tensor<1x1x1x2xf32>} : () -> (tensor<*xf32>)
  
  // Define the bias tensor
  %bias, %ctl_bias = Const device("/device:CPU:0") name("bias") {dtype = f32, value = dense<[0.531091094, -0.719168067]> : tensor<2xf32>} : () -> (tensor<*xf32>)

  // Perform convolution operation
  %conv2d, %ctl_conv = Conv2D(%input_tensor, %filter) device("/device:CPU:0") name("conv2d") {T = f32, data_format = "NHWC", dilations = [1, 1, 1, 1], explicit_paddings = [], padding = "SAME", strides = [1, 1, 1, 1], use_cudnn_on_gpu = true} : (tensor<*xf32>, tensor<*xf32>) -> (tensor<*xf32>)
  
  // Squeeze the output of the convolution
  %squeezed, %ctl_squeeze = Squeeze(%conv2d) device("/device:CPU:0") name("squeezed") {T = f32, squeeze_dims = [2]} : (tensor<*xf32>) -> (tensor<*xf32>)
  
  // Add bias to the squeezed tensor
  %bias_added, %ctl_bias_add = BiasAdd(%squeezed, %bias) device("/device:CPU:0") name("bias_added") {T = f32, data_format = "NHWC"} : (tensor<*xf32>, tensor<*xf32>) -> (tensor<*xf32>)
  
  return
}
```

Renaming Operations:

- Changed `%Placeholder` to `%input_tensor`, `%Const` to `%filter`, and `%Const_1` to `%bias` for clarity. Clear names help in understanding the role of each tensor.

Simplifying Names:

- Adjusted operation names (`Conv2D` to `conv2d`, `Squeeze` to `squeezed`, and `BiasAdd` to `bias_added`) to be more descriptive and aligned with common TensorFlow naming conventions.

Consistent Naming:

- Used consistent naming conventions for tensors and operations to enhance readability. For instance, `%ctl` names are now aligned with their corresponding tensors.

Maintained Device and Type Annotations:

- Ensured that the device and type annotations remain intact for clarity on the computational resource and data type.

**Pull request closed without merging**